js_modules/dagster-ui/packages/ui-core/src/asset-graph/AssetEdges.tsx

```diff
@@ -13,8 +15,121 @@ interface AssetEdgesProps {
+
+    //https://stackoverflow.com/a/20925869/1162881
+    function doesViewportContainEdge(
+      edge: {from: {x: number; y: number}; to: {x: number; y: number}},
+      viewportRect: {top: number; left: number; right: number; bottom: number},
+    ) {
+      return (
+        isOverlapping1D(
+          Math.max(edge.from.x, edge.to.x),
+          Math.max(viewportRect.left, viewportRect.right),
+          Math.min(edge.from.x, edge.to.x),
+          Math.min(viewportRect.left, viewportRect.right),
+        ) &&
+        isOverlapping1D(
+          Math.max(edge.from.y, edge.to.y),
+          Math.max(viewportRect.top, viewportRect.bottom),
+          Math.min(edge.from.y, edge.to.y),
+          Math.min(viewportRect.top, viewportRect.bottom),
+        )
+      );
+    }
+
+    function isOverlapping1D(xmax1: number, xmax2: number, xmin1: number, xmin2: number) {
+      return xmax1 >= xmin2 && xmax2 >= xmin1;
+    }
 
+    function doesViewportContainPoint(
+      point: {x: number; y: number},
+      viewportRect: {top: number; left: number; right: number; bottom: number},
+    ) {
+      return (
+        point.x >= viewportRect.left &&
+        point.x <= viewportRect.right &&
+        point.y >= viewportRect.top &&
+        point.y <= viewportRect.bottom
+      );
+    }
+
+    // Note: we render the highlighted edges twice, but it's so that the first item with
+    // all the edges in it can remain memoized.
+
+    // Round to the nearest 100px to avoid recalculating edges too often (could be an array of 1,000+ edges)
+    const left = Math.floor(viewportRect.left / 100) * 100;
+    const right = Math.ceil(viewportRect.right / 100) * 100;
+    const top = Math.floor(viewportRect.top / 100) * 100;
+    const bottom = Math.ceil(viewportRect.bottom / 100) * 100;
+
+    const edgesToShow = (() => {
+      const rectRounded = {left, right, top, bottom};
+      const intersectedEdges = edges.filter((edge) => doesViewportContainEdge(edge, rectRounded));
+      if (intersectedEdges.length <= 10) {
+        return intersectedEdges;
+      }
+      const visibleToFromEdges = intersectedEdges.filter(
+        (edge) =>
+          doesViewportContainPoint(edge.from, rectRounded) ||
+          doesViewportContainPoint(edge.to, rectRounded),
+      );
+      if (visibleToFromEdges.length < 50) {
+        return visibleToFromEdges;
+      }
+      const center = {
+        x: (rectRounded.left + rectRounded.right) / 2,
+        y: (rectRounded.top + rectRounded.bottom) / 2,
+      };
+      const edgesWithDistance = visibleToFromEdges.map((edge) => ({
+        edge,
+        distance: Math.min(
+          Math.sqrt(Math.pow(edge.from.x - center.x, 2) + Math.pow(edge.from.y - center.y, 2)),
+          Math.sqrt(Math.pow(edge.to.x - center.x, 2) + Math.pow(edge.to.y - center.y, 2)),
+        ),
+      }));
+      edgesWithDistance.sort((a, b) => a.distance - b.distance);
+      return edgesWithDistance.slice(0, MAX_EDGES).map((item) => item.edge);
```

Comment:
Ahh I'd never noticed this second layer of filtering in here!

Because we only use the distance values for sorting, and the distances and the square of the distances will sort the same, we'd get a nice speed boost by not computing the `Math.sqrt` of each squared distance. I think it's super expensive relative to the X*X part.

Reply:
Ah yeah good catch!